assembly/src/tokens/stream.rs

```diff
@@ -23,59 +20,96 @@ impl<'a> TokenStream<'a> {
-                        // which is followed by a blank line.
-                        return Err(ParsingError::dangling_procedure_comment(line_number));
-                    }
+
+        // iterate the lines
+        for (n, info) in LinesStream::from(source).enumerate() {
+            let line_number = info.line_number() as usize;
+
+            // fetch the line contents
+            let mut line = match info.contents() {
+                // postfix docs are always dangling
+                _ if info.postfix_docs().is_some() => {
+                    return Err(ParsingError::dangling_procedure_comment(line_number));
+                }
+
+                // first dangling is mod comment; otherwise, if contents empty & docs, then dangling
+                None if n != 0 && !info.docs().is_empty() => {
+                    return Err(ParsingError::dangling_procedure_comment(line_number));
                 }
-            } else {
-                let mut line_tokens = line
-                    .split_whitespace()
-                    .take_while(|&token| !token.starts_with(LINE_COMMENT_PREFIX))
-                    .collect::<Vec<_>>();
-
-                if !comment_builder.is_empty() {
-                    // procedure comment should always be followed by a procedure token
-                    debug_assert!(!line_tokens.is_empty());
-                    let token = line_tokens[0];
-                    if token.starts_with(Token::EXPORT) || token.starts_with(Token::PROC) {
-                        proc_comments.insert(tokens.len(), comment_builder.take_comment());
-                    } else {
-                        return Err(ParsingError::dangling_procedure_comment(line_number));
+
+                // only procs allowed for doc comment
+                Some(line)
+                    if n != 0
+                        && !info.docs().is_empty()
+                        && !(line.trim().starts_with(Token::EXPORT)
+                            || line.trim().starts_with(Token::PROC)) =>
+                {
+                    return Err(ParsingError::dangling_procedure_comment(line_number));
+                }
+
+                // line found
+                Some(line) => line,
+
+                // no contents available & no exception to throw
+                None => continue,
+            };
+
+            // iterate columns from the first
+            let mut _column = 1;
+            while !line.trim().is_empty() {
+                // eat preceding empty cols
+                if let Some(n) = line.find(|c: char| !c.is_whitespace()) {
+                    _column += n;
+                    line = line.split_at(n).1;
+                }
+
+                // extract token
+                let token = match line.find(|c: char| c.is_whitespace()) {
+                    Some(n) => {
+                        let (token, remainder) = line.split_at(n);
+                        line = remainder;
+                        token
                     }
```

Comment:
If we do something similar to my previous comment, changes in this part would be minimal as well. For example, we could use something like:
```Rust
// fetch all tokens
for line_info in LineStream::from(source) {
    match line_info.contents() {
        Some(line) => {
            if line.starts_with(Token::EXPORT) || line.starts_with(Token::PROC) {
                let doc_comment = build_comment(line_info.docs());
                proc_comments.insert(tokens.len(), doc_comment);
            } else if !line_info.docs().is_empty() {
                return Err(ParsingError::dangling_procedure_comment(
                    line_info.line_number() as usize,
                ));
            }

            let mut line_tokens = line
                .split_whitespace()
                .take_while(|&token| !token.starts_with(Token::LINE_COMMENT_PREFIX))
                .collect::<Vec<_>>();

            tokens.append(&mut line_tokens);
            lines.resize(tokens.len(), line_info.line_number() as usize);
        }

        None if tokens.is_empty() => {
            module_comment = build_comment(line_info.docs());
        }

        None => {
            return Err(ParsingError::dangling_procedure_comment(
                line_info.line_number() as usize
            ))
        }
    }
}
```